lib/socket.js

```diff
@@ -62,20 +61,64 @@ class Transport extends EventEmitter {
   //
   _onSocketData(chunk) {
     const packets = [];
-    this._buffer += chunk;
+    let parsedLength;
+    let newPartLength;
+
+    if (this._buffer) {
+      newPartLength = chunk.indexOf(0);
+      if (newPartLength === -1) {
+        this._buffer = Buffer.concat(
```

Comment:
@belochub surely, empirical way will be better, I just suggested a starting point to bench different values. Also the way I propose should only kick in if we receive half-message and in case we receive full message we should not copy it to the intermediate buffer.
Regarding the idea: now when we receive first half-message we concat it with null(initial state of class buffer), upon each next half-message we receive we will copy current class buffer and received data to the new bigger buffer with the size exactly the sum of the previous' buffer sizes. Hence every time we receive a message we will copy every data we've received so far to the new buffer and then copy the received data. Please correct me if I see it wrong.
With the way I propose to do it we will copy the first half-message to the class buffer (which will always exist of some size), then upon receiving consequent half-messages we will append them to that class buffer hence only coping the received data. This way we will only copy our accumulated data if there is not enough space in the class buffer (and we will grow it more that just the size of received data), hence avoiding a lot of copying in case of very fragmented data.